### PR TITLE
option to gather messages even without defaultMessage attribute

### LINKED ISSUE #43: Required defaultMessage
What is the reason behind requiring the "defaultMessage" field when extracting the strings?
React-Intl can work perfectly fine without this attribute.

Would it be feasible to have an option/parameter for plugin not to require it if specified?
I could do it, and create a PR.

### PR DESCRIPTION
**Pull request by bogzio**
Should fix #43

### CODE REVIEW

**Comment by krzkaczor** [comment]
This is very useful PR. Documentation doesnt mention the fact that you need to provide defaultMessage (which I think is not a great idea since I want to store all my translactions in json files).

**Comment by redonkulus** [comment]
@bziorkiewicz want to add a test for the new option?

**Comment by SleepWalker** [comment]
Btw, I have a similar #140 PR with tests included. But there are some bonus options :D and no docs.

P.s. this PR will break the logic, that clean ups the code from descriptions.

**Comment by langpavel** [comment]
See also https://github.com/yahoo/babel-plugin-react-intl/pull/137#issuecomment-384022388

**Comment by redonkulus** [comment]
@SleepWalker @langpavel I'm not familiar with the code, so I'm offering to help since there seems to be 3 PR's that do the same thing?

I just want to make sure we do not break any existing functionality, we have proper documentation and tests. I would prefer to consolidate the various PRs perhaps in favor of #140.

**Comment by langpavel** [comment]
@redonkulus Yeah, there is too much forks and much more features around the github, seems like `babel-plugin-react-intl` is **really *carefully* maintained** (but this looks like not developed…)
Of course, I have all tests passed too and even more, I write new tests for new features in #137.

BTW @bziorkiewicz, please, what you think about #137, can it solve your problem, can you review it? I think I did better job :grin:

File: src/index.js [hunk 1]
```diff
@@ -251,7 +257,7 @@ export default function ({types: t}) {
                     // `<FormattedMessage id={dynamicId} />`, because it will be
                     // skipped here and extracted elsewhere. The descriptor will
                     // be extracted only if a `defaultMessage` prop exists.
-                    if (descriptor.defaultMessage) {
+                    if (descriptor.defaultMessage || opts.optionalDefaultMessages) {
```
Comment (rasendubi) [inline src/index.js]: This breaks the plugin when spreads are used. e.g., `defineMessages` test case fails:
```
import React, {Component} from 'react';
import {defineMessages, FormattedMessage} from 'react-intl';

const msgs = defineMessages({
    header: {
        id: 'foo.bar.baz',
        defaultMessage: 'Hello World!',
        description: 'The default message',
    },
    content: {
        id: 'foo.bar.biff',
        defaultMessage: 'Hello Nurse!',
        description: 'Another message',
    },
});

export default class Foo extends Component {
    render() {
        return (
            <div>
                <h1><FormattedMessage {...msgs.header}/></h1>
                <p><FormattedMessage {...msgs.content}/></p>
            </div>
        );
    }
}
```

A proper fix could be checking for `id` presence instead. See: https://github.com/Doctoright/babel-plugin-react-intl/commit/2e13607b63b3ae4aa911a246992ef2c8a4f299af

**Comment by Keroxo** [comment]
What's the status about this? This PR would be relevant for me too.

**Comment by rasendubi** [comment]
@Keroxo there seems to be no movement and there is little hope it will be.

I have applied the patch with my fix mentioned above and published it as [@doctoright/babel-plugin-react-intl](https://www.npmjs.com/package/@doctoright/babel-plugin-react-intl).

**Comment by Keroxo** [comment]
@rasendubi okay. sad to hear that. Thanks for your fast response

**Comment by longlho** [comment]
Can you add a test please?

**Review by longlho — CHANGES_REQUESTED** [review]
Can u add tests please?

**Comment by longlho** [comment]
close due to stale

**Pull request closed without merging**